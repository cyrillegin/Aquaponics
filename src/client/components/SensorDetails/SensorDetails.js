import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {withStyles} from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';

const styles = {
  root: {
    margin: '16px',
  },
  container: {
    display: 'flex',
    flexWrap: 'wrap',
  },
  textField: {
    margin: '20px',
    width: '250px',
  },
  paper: {
    padding: '20px',
  },
};

export class SensorDetails extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    updateSensor: PropTypes.func.isRequired,
    deleteSensor: PropTypes.func.isRequired,
    sensor: PropTypes.shape({
      uuid: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      created: PropTypes.number.isRequired,
      modified: PropTypes.number.isRequired,
      poller: PropTypes.string.isRequired,
      pin: PropTypes.string.isRequired,
      pollRate: PropTypes.number.isRequired,
      status: PropTypes.string.isRequired,
      station: PropTypes.string.isRequired,
      description: PropTypes.string,
      coefficients: PropTypes.string,
      units: PropTypes.string,
      endpoint: PropTypes.string,
    }),
  };

  static defaultProps = {
    sensor: null,
  }

  state = {
    uuid: '',
    name: '',
    description: '',
    coefficients: '',
    station: '',
    poller: '',
    pin: '',
    units: '',
    endpoint: '',
    pollRate: '',
    status: '',
    localIP: '',
    loading: true,
    deleteDialogIsOpen: false,
  }

  handleChange = name => event => {
    this.setState({
      [name]: event.target.value,
    });
  };

  componentDidUpdate(prev) {
    if ((prev.sensor === null && this.props.sensor !== null) ||
    (prev.sensor !== null && prev.sensor.uuid !== this.props.sensor.uuid)) {
      this.setState({
        loading: false,
        uuid: this.props.sensor.uuid || '',
        created: this.props.sensor.created || '',
        modified: this.props.sensor.modified || '',
        name: this.props.sensor.name || '',
        description: this.props.sensor.description || '',
        coefficients: this.props.sensor.coefficients || '',
        station: this.props.sensor.station || '',
        poller: this.props.sensor.poller || '',
        pin: this.props.sensor.pin || '',
        units: this.props.sensor.units || '',
        endpoint: this.props.sensor.endpoint || '',
        pollRate: this.props.sensor.pollRate || '',
        status: this.props.sensor.status || '',
      });
    }
  }

  componentDidMount() {
    fetch('/getIP')
      .then(response => response.json())
      .then((data) => {
        this.setState({
          localIP: data.localIP,
        });
      });
  }

  render() {
    const updateSensor = () => {
      this.props.updateSensor({
        uuid: this.state.uuid,
        name: this.state.name,
        description: this.state.description,
        coefficients: this.state.coefficients,
        station: this.state.station,
        poller: this.state.poller,
        pin: this.state.pin,
        units: this.state.units,
        endpoint: this.state.endpoint,
        pollRate: this.state.pollRate,
        status: this.state.status,
      });
    };

    const openDeleteSensorDialog = () => {
      this.setState({
        deleteDialogIsOpen: true,
      });
    };

    const closeDeleteSensorDialog = () => {
      this.setState({
        deleteDialogIsOpen: false,
      });
    };

    const handleDeleteSensor = () => {
      this.props.deleteSensor().then((res) => {
        closeDeleteSensorDialog();
      });
    };

    if (this.state.loading) {
      return (<div />);
    } else if (this.props.sensor.uuid === '' || this.props.sensor.uuid === null) {
      return (<div />);
    }
    return (
      <div className={this.props.classes.root}>

        <Dialog
          open={this.state.deleteDialogIsOpen}
          onClose={closeDeleteSensorDialog}
          aria-labelledby="alert-dialog-title"
          aria-describedby="alert-dialog-description"
        >
          <DialogTitle id="alert-dialog-title">{'Delete Sensor?'}</DialogTitle>
          <DialogContent>
            <DialogContentText id="alert-dialog-description">
              Are you sure you want to delete this sensor?
            </DialogContentText>
          </DialogContent>
          <DialogActions>
            <Button onClick={handleDeleteSensor} color="primary">
              Yes
            </Button>
            <Button onClick={closeDeleteSensorDialog} color="primary" autoFocus>
              No
            </Button>
          </DialogActions>
        </Dialog>


        <Paper className={this.props.classes.paper} elevation={4}>
          <Typography variant="headline" component="h3">
            Sensor Details
          </Typography>
          <hr />
          <div className={this.props.classes.container}>
            <TextField
              id="uuid"
              label="UUID"
              className={this.props.classes.textField}
              value={this.state.uuid}
              margin="normal"
              disabled
            />
            <TextField
              id="created"
              label="Created"
              className={this.props.classes.textField}
              value={new Date(this.state.created).toGMTString()}
              margin="normal"
              disabled
            />
            <TextField
              id="modified"
              label="Last Modified"
              className={this.props.classes.textField}
              value={new Date(this.state.modified).toGMTString()}
              margin="normal"
              disabled
            />
            <TextField
              id="localip"
              label="Local IP Address"
              className={this.props.classes.textField}
              value={this.state.localIP}
              margin="normal"
              disabled
            />
          </div>

          <form className={this.props.classes.container} noValidate autoComplete="off">
            <TextField
              id="name"
              label="Name"
              onChange={this.handleChange('name')}
              className={this.props.classes.textField}
              value={this.state.name}
              margin="normal"
            />
            <TextField
              id="description"
              label="Description"
              onChange={this.handleChange('description')}
              className={this.props.classes.textField}
              value={this.state.description}
              margin="normal"
            />
            <TextField
              id="coefficients"
              label="Coefficients"
              onChange={this.handleChange('coefficients')}
              className={this.props.classes.textField}
              value={this.state.coefficients}
              margin="normal"
            />
            <TextField
              id="station"
              label="Station"
              onChange={this.handleChange('station')}
              className={this.props.classes.textField}
              value={this.state.station}
              margin="normal"
            />
            <TextField
              id="pollRate"
              label="Poll Rate"
              onChange={this.handleChange('pollRate')}
              className={this.props.classes.textField}
              value={this.state.pollRate}
              margin="normal"
            />
            <TextField
              id="poller"
              label="Poller"
              onChange={this.handleChange('poller')}
              className={this.props.classes.textField}
              value={this.state.poller}
              margin="normal"
            />
            <TextField
              id="pin"
              label="Pin"
              onChange={this.handleChange('pin')}
              className={this.props.classes.textField}
              value={this.state.pin}
              margin="normal"
            />
            <TextField
              id="units"
              label="Units"
              onChange={this.handleChange('units')}
              className={this.props.classes.textField}
              value={this.state.units}
              margin="normal"
            />
            <TextField
              id="endpoint"
              label={'Endpoint'}
              onChange={this.handleChange('endpoint')}
              className={this.props.classes.textField}
              value={this.state.endpoint}
              margin="normal"
            />
            <TextField
              id="status"
              label={'Status'}
              onChange={this.handleChange('status')}
              className={this.props.classes.textField}
              value={this.state.status}
              margin="normal"
            />
            <Button className={this.props.classes.button} onClick={updateSensor}>Submit Changes</Button>
            <Button className={this.props.classes.button} onClick={openDeleteSensorDialog}>Delete Sensor</Button>
          </form>
        </Paper>
      </div>
    );
  }
}

export default withStyles(styles)(SensorDetails);
