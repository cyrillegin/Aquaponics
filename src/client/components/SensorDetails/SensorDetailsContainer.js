import {compose, mapProps} from 'recompose';
import SensorDetails from './SensorDetails';

export default compose(
  mapProps((ownProps) => {
    return {
      sensor: ownProps.sensor,
      updateSensor: (sensor) => {
        return new Promise((res, rej) => {
          fetch('/api/sensor', {
            method: 'PUT',
            headers: {
              'Accept': 'application/json',
              'Content-Type': 'application/json',
            },
            body: JSON.stringify({...sensor}),
          })
            .then(response => response.json())
            .then((data) => {
              res({'data': 'data'});
              window.location.reload();
            });
        });
      },
      deleteSensor: () => {
        return new Promise((res, rej) => {
          fetch('/api/sensor/?sensor=' + ownProps.sensor.uuid, {
            method: 'DELETE',
            headers: {
              'Accept': 'application/json',
              'Content-Type': 'application/json',
            },
            body: JSON.stringify({...ownProps.sensor}),
          })
            .then(response => response.json())
            .then((data) => {
              ownProps.history.push('/');
              res({'data': data});
              window.location.reload();
            });
        });
      },
      addEntry: (value) => {
        return fetch('/api/reading', {
          method: 'POST',
          headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({uuid: ownProps.sensor.uuid, value: value, sensor: ownProps.sensor.name}),
        });
      },
    };
  }),
)(SensorDetails);
