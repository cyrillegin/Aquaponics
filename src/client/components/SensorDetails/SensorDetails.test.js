import { mount } from 'enzyme';
import React from 'react';
import { act } from '@testing-library/react';
import SensorDetails from './SensorDetails';

jest.useFakeTimers();

describe('SensorDetails', () => {
  let fetch;
  let date;

  beforeEach(() => {
    fetch = global.fetch;
    global.fetch = jest.fn(
      () => new Promise(res => res({ json: () => new Promise(inner => inner({})) })),
    );
    date = global.Date;
    global.Date = class {
      getDate = () => '1';

      getMonth = () => '1';

      getFullYear = () => '1';

      getHours = () => '1';

      getMinutes = () => '1';

      static now() {
        return '1';
      }
    };
  });

  afterEach(() => {
    global.fetch = fetch;
    global.Date = date;
  });

  it('should render a snap shot', () => {
    act(() => {
      const wrapper = mount(
        <SensorDetails
          sensor={{
            id: 1,
            name: 'test',
            health: 'healthy',
            stationId: 1,
            stationName: 'test',
            lastReading: {
              timestamp: '2020-10-04T20:59:30.176Z',
              value: 1,
            },
            actions: [{ condition: 'test', interval: 'test', action: 'test', id: 1 }],
          }}
        />,
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  it('should test functionality', () => {
    act(() => {
      const wrapper = mount(
        <SensorDetails
          sensor={{
            id: 1,
            name: 'test',
            health: 'healthy',
            stationId: 1,
            stationName: 'test',
            lastReading: {
              timestamp: '2020-10-04T20:59:30.176Z',
              value: 1,
            },
            actions: [{ condition: 'test', interval: 'test', action: 'test', id: 1 }],
          }}
        />,
      );

      wrapper.find('button').forEach(button => {
        button.simulate('click');
        jest.runAllTimers();
        expect(wrapper.html()).toMatchSnapshot();
      });
    });
  });
});
