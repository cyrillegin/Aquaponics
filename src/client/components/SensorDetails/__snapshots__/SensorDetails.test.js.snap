// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SensorDetails should render a snap shot 1`] = `ReactWrapper {}`;

exports[`SensorDetails should test functionality 1`] = `"<div class=\\"css-1b5h7m0\\"><div class=\\"grid\\"><div class=\\"item\\">Station:<span>test</span></div><div class=\\"item\\">Sensor:<span>test</span></div><div class=\\"item\\">Last Reading Time:<span>1-11-1 1:1</span></div><div class=\\"item\\">Last Reading Value:<span>1</span></div><div class=\\"item\\">Update name:<input name=\\"sensorName\\" value=\\"test\\"></div><div class=\\"item\\">Description<input name=\\"sensorDescription\\" value=\\"\\"></div><div class=\\"item\\">Coefs:<input name=\\"sensorCoefs\\" value=\\"\\"></div><div class=\\"item\\"><button type=\\"button\\">save</button><span>Name must be something</span></div></div><div class=\\"actions-title\\">Actions</div><div class=\\"actions-grid\\"><div class=\\"actions-item\\">Value</div><div class=\\"actions-item\\">Condition</div><div class=\\"actions-item\\">action</div><div class=\\"actions-item\\">interval</div><div class=\\"actions-item\\">meta?</div></div></div>"`;
